Dataset class additions

ehanson8 opened the pull request:
Adds classes and attributes to expand the `Dataset` class and modifies the `asdict` method to account for these new attributes. Precursor to finishing the WHOAS ingest

ehanson8 committed:
Dataset class additions

hakbailey commented on hoard/models.py:
```diff
@@ -19,22 +19,95 @@ class Contact:
     description: List[Description]
     subjects: List[str]
     title: str
+    alternateUrl: Optional[str] = None
```
This is totally my bad because it was incorrect in the data mapping spreadsheet, but this field should be named `alternativeURL`.

hakbailey commented on hoard/models.py:
```diff
@@ -19,22 +19,95 @@ class Contact:
     def asdict(self) -> dict:
         fields = [primitive(self.title, "title")]
+        if self.alternateUrl is not None:
+            fields = [primitive(self.alternateUrl, "alternateUrl")]
```
This should append to `fields` rather than replace it. There are many other attributes where this has been done and they all need to be fixed.

hakbailey commented on hoard/models.py:
```diff
@@ -58,10 +131,79 @@ def asdict(self) -> dict:
+                compound(
+                    self.distributors,
+                    "distributor",
+                    ["distributorName", "distributorType"],
```
This should be `["distributorName", "distributorURL"]` as there is no `distributorType` in our model.

hakbailey commented on hoard/models.py:
```diff
@@ -19,22 +19,95 @@ class Contact:
+    distributionDate: Optional[str] = None
+    distributors: Optional[List[Distributor]] = None
+    grantNumbers: Optional[List[GrantNumber]] = None
+    keywordValue: Optional[str] = None
```
I would change this to `keywords`, and it should be a list of strings (we will definitely have multiple keywords for a given record). See comment below on the `keyword` field in Dataverse for more on why.

hakbailey commented on hoard/models.py:
```diff
@@ -58,10 +131,79 @@ def asdict(self) -> dict:
+                )
+            )
+        if self.keywordValue is not None:
+            fields = [primitive(self.keywordValue, "keywordValue")]
```
`keyword` is a compound field in Dataverse, but we are only using one of the subfields. So this should be:
```
if self.keywords is not None: # see above comment on changing this attribute name and type
    fields.append(
        compound(
            self.keywords,
                "keyword",
                ["keywordValue"],
            )
        )
```

hakbailey commented on hoard/models.py:
```diff
@@ -19,22 +19,95 @@ class Contact:
+    distributors: Optional[List[Distributor]] = None
+    grantNumbers: Optional[List[GrantNumber]] = None
+    keywordValue: Optional[str] = None
+    kindOfData: Optional[str] = None
```
It turns out `kindOfData` can have multiple values (again my bad, I didn't notice that until just now) so this should also be a list of strings.

hakbailey commented on hoard/models.py:
```diff
@@ -58,10 +131,79 @@ def asdict(self) -> dict:
+        if self.keywordValue is not None:
+            fields = [primitive(self.keywordValue, "keywordValue")]
+        if self.kindOfData is not None:
+            fields = [primitive(self.kindOfData, "kindOfData")]
```
Annoyingly, although `kindOfData` is a primitive field in Dataverse, it does allow multiple values. So we should just create the dict for this manually like so:
```
fields.append(
    {
        "typeName": "kindOfData",
        "multiple": True,
        "typeClass": "primitive",
        "value":  self.kindOfData,
    }
)

hakbailey reviewed (changes requested):
In addition to the specific comments below, the tests need to be updated to check that we can create a Dataset with all possible fields in addition to just the required ones, and that the `asdict()` method returns a dict in the correct Dataverse structure. You can use [this sample json file](http://guides.dataverse.org/en/latest/_downloads/dataset-create-new-all-default-fields.json) provide by Dataverse as a starting point, but you'll need to edit it to remove the fields we aren't using in our model, as well as adding the language field which for some reason isn't included.

ehanson8 committed:
PR updates

Major changes:
- per slack discussion: temp fix to set some attributes as typeClass "controlledVocabulary"
- fix for novelty of series attribute
-create full and minimal JSON fixtures for testing and updating tests accordingly

ehanson8 committed:
Update models.py

hakbailey reviewed (approved):
👍
